### PR TITLE
Add comments option to playground

### LINKED ISSUE #6509: Playground option for "print comments"
In https://prettier.io/playground/ you can toggle an option to display the comments. It'd be nice to add that to https://play.ruff.rs/.

I mentioned this on Discord and didn't want to lose track of it. I can't claim this yet, but I'm happy to come back here to grab it if I can.

### PR DESCRIPTION
**Pull request by cnpryer**
Closes #6509

This PR adds a comments option to the playground by implementing a public `pretty_comments` function in `ruff_python_formatter`.

image

TODO:
- [x] Public `print_comments` or similar function [(comment)](https://github.com/astral-sh/ruff/issues/6509#issuecomment-1694546839)

### CODE REVIEW

File: playground/src/Editor/Icons.tsx [hunk 1]
```diff
@@ -124,3 +124,9 @@ export function FormatterIRIcon() {
     </svg>
   );
 }
+  // TODO(cnpryer)
+  export function CommentsIcon() {
+    return (
+      <svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg" fill="#ffffff"><g id="SVGRepo_bgCarrier" stroke-width="0"></g><g id="SVGRepo_tracerCarrier" stroke-linecap="round" stroke-linejoin="round"></g><g id="SVGRepo_iconCarrier"> <title></title> <g id="Complete"> <g id="bubble-square"> <path d="M7.7,18.3H19.4a2.1,2.1,0,0,0,2.1-2.1V4.6a2.1,2.1,0,0,0-2.1-2.1H4.6A2.1,2.1,0,0,0,2.5,4.6V21.5Z" fill="none" stroke="#ffffff" stroke-linecap="round" stroke-linejoin="round" stroke-width="1"></path> </g> </g> </g></svg>
```
Comment (cnpryer) [inline playground/src/Editor/Icons.tsx:130]: I have to look into SVG codes

File: crates/ruff_python_formatter/src/comments/debug.rs [hunk 2]
```diff
@@ -37,7 +37,7 @@ impl Debug for DebugComment<'_> {
 }
 
 /// Pretty-printed debug representation of [`Comments`].
-pub(crate) struct DebugComments<'a> {
```
Comment (cnpryer) [inline crates/ruff_python_formatter/src/comments/debug.rs:40]: Ignore this

File: crates/ruff_python_formatter/src/comments/mod.rs [hunk 3]
```diff
@@ -282,7 +282,7 @@ type CommentsMap<'a> = MultiMap<NodeRefEqualityKey<'a>, SourceComment>;
 ///
 /// Cloning `comments` is cheap as it only involves bumping a reference counter.
 #[derive(Debug, Clone, Default)]
-pub(crate) struct Comments<'a> {
```
Comment (cnpryer) [inline crates/ruff_python_formatter/src/comments/mod.rs:285]: Ignore this

File: crates/ruff_python_formatter/src/comments/mod.rs [hunk 4]
```diff
@@ -463,7 +463,7 @@ impl<'a> Comments<'a> {
     }
 
     /// Returns an object that implements [Debug] for nicely printing the [`Comments`].
-    pub(crate) fn debug(&'a self, source_code: SourceCode<'a>) -> DebugComments<'a> {
```
Comment (cnpryer) [inline crates/ruff_python_formatter/src/comments/mod.rs:466]: Ignore this

File: crates/ruff_python_formatter/src/context.rs [hunk 5]
```diff
@@ -40,7 +40,7 @@ impl<'a> PyFormatContext<'a> {
         self.node_level
     }
 
-    pub(crate) fn comments(&self) -> &Comments<'a> {
```
Comment (cnpryer) [inline crates/ruff_python_formatter/src/context.rs:43]: Ignore this

File: crates/ruff_wasm/Cargo.toml [hunk 6]
```diff
@@ -22,6 +22,7 @@ ruff = { path = "../ruff" }
 ruff_diagnostics = { path = "../ruff_diagnostics" }
 ruff_python_ast = { path = "../ruff_python_ast" }
 ruff_python_codegen = { path = "../ruff_python_codegen" }
+ruff_formatter = { path = "../ruff_formatter" }
```
Comment (cnpryer) [inline crates/ruff_wasm/Cargo.toml:25]: Ignore this

File: crates/ruff_wasm/src/lib.rs [hunk 7]
```diff
@@ -22,6 +22,7 @@ use ruff::settings::configuration::Configuration;
 use ruff::settings::options::Options;
 use ruff::settings::types::PythonVersion;
 use ruff::settings::{defaults, flags, Settings};
+use ruff_formatter::SourceCode;
```
Comment (cnpryer) [inline crates/ruff_wasm/src/lib.rs:25]: Ignore this

File: playground/src/Editor/setupMonaco.tsx [hunk 8]
```diff
@@ -620,6 +621,62 @@ function defineRustPythonAstLanguage(monaco: Monaco) {
   });
 }
 
+// Modeled after 'RustPythonAst'
+function defineCommentsLanguage(monaco: Monaco) {
```
Comment (cnpryer) [inline playground/src/Editor/setupMonaco.tsx:625]: I tried looking into highlighting for ``code snippets``, but couldn't figure it out with a first-pass.

File: playground/src/Editor/Icons.tsx (hunk 1 above)
Comment (cnpryer) [reply playground/src/Editor/Icons.tsx]: IIUC it's just a type of encoding? I'll revisit if it's preferred.

File: crates/ruff_wasm/src/lib.rs [hunk 9]
```diff
@@ -294,6 +294,25 @@ impl Workspace {
...
+        let tokens: Vec<_> = ruff_python_parser::lexer::lex(contents, Mode::Module).collect();
+        let mut comment_ranges = CommentRangesBuilder::default();
+
+        for (token, range) in tokens.iter().flatten() {
+            comment_ranges.visit_token(token, *range);
+        }
+        let comment_ranges = comment_ranges.finish();
+        let module = parse_tokens(tokens, Mode::Module, ".").map_err(into_error)?;
+
+        // TODO(konstin): Add an options for py/pyi to the UI (2/2)
```
Comment (cnpryer) [inline crates/ruff_wasm/src/lib.rs:307]: Left this here intentionally

File: playground/src/Editor/SecondaryPanel.tsx [hunk 10]
```diff
@@ -64,6 +65,10 @@ function Content({
           case "FIR":
             language = "fir";
             break;
+          
+          case "Comments":
+            language = "Comments";
```
Comment (cnpryer) [inline playground/src/Editor/SecondaryPanel.tsx:70]: Could probably call this "Formatted Comments"

File: crates/ruff_python_formatter/src/lib.rs [hunk 11]
```diff
@@ -167,6 +167,21 @@ pub fn format_node<'a>(
...
+        .context()
+        .comments()
+        .debug(SourceCode::new(source));
+
+    // When comments are empty we'd display an empty map '{}'
+    if comments.empty() {
+        String::new()
+    } else {
+        std::format!("{comments:#?}")
+    }
```
Comment (cnpryer) [inline crates/ruff_python_formatter/src/lib.rs:182]: image

**Review by MichaReiser — APPROVED** [review]
This is so cool. Love it

File: crates/ruff_python_formatter/src/lib.rs (hunk 11 above)
Comment (cnpryer) [reply crates/ruff_python_formatter/src/lib.rs]: That'd be the expected output as-is (with no condition on empty comments). It was a nit of mine so I decided to print nothing if comments were 'empty'.

**Comment by MichaReiser** [comment]
This is now deployed to https://play.ruff.rs

**Pull request merged**